### ISSUE TITLE
[jvm-packages] train model on spark is not stable

### ISSUE THREAD

**Issue by wangzhilong** [dmlc/xgboost#2573]
I train a model on spark with train_data and calculate auc with test_data.  I repeated the action , but every time, the auc is different.  sometimes it is up to 0.79, and sometimes  it is as low as 0.76.

train code:
`val paramMap = List(`
 `     "booster" -> "gbtree",`
 `     "eta" -> 0.5,`
  `    "max_depth" -> 4,`
  `    "tree_method" -> "exact",`
  `    "save_period" -> 0,`
   `   "eval_metric" -> "auc",`
 `    "objective" -> "binary:logistic").toMap`
`val round = 100`
`val nWorkers = 32`
`model = XGBoost.trainWithRDD(trainData, paramMap, round, nWorkers = nWorkers, useExternalMemory = true)`

auc code:
 `val predict = model.transform(testData)`
 `val evaluator = new BinaryClassificationEvaluator().setLabelCol("label").
      setRawPredictionCol("probabilities").setMetricName("areaUnderROC")`
  `val auc = evaluator.evaluate(predict)`

**Comment by CodingCat** [comment]
how many samples do you have?

**Comment by wangzhilong** [comment]
train_data: 6515508x43
test_data:  1729773x43

**Comment by khotilov** [comment]
@wangzhilong  What is your positive class rate, and do all workers get chunks of data with similar properties?

Keep in mind that AUC is calculated approximately in distributed settings, by averaging individual workers' AUC's.

**Comment by wangzhilong** [comment]
@khotilov Thanks for your reply。 I trained model in spark and calculated auc in local, not in  distributed setting.

`data = xgb.DMatrix(data_path)`
`bst = xgb.Booster({'nthrea': 4})`
`bst.load_model(model_path)`
`labels = data.get_label()`
`pred = bst.predict(data)`
`print metrics.roc_auc_score(labels, pred)`

**Comment by CodingCat** [comment]
are you using libsvm format?

**Comment by wangzhilong** [comment]
@CodingCat I'm using libsvm format, but i process the data index, convert one-based to zero-based.

**Comment by wangzhilong** [comment]
@CodingCat 
today, i changed params: round=100 to round=30, and trained 4 times. on test_data the auc is
stabled at 0.79. may round = 100, overfitting.

i don't know why train on spark is not stable when rount=100.
if i train in local, same params can get same model, every time.

**Comment by CodingCat** [comment]
when you say train in local, it is spark local model, or single thread training?

**Comment by wangzhilong** [comment]
use "./xgboost train.conf",  not spark local model. @CodingCat

**Comment by CodingCat** [comment]
then I would say it is expected when you separate training set into multiple partitions,

the split point proposed in each partition is based on the data *within* that partition, bringing some accuracy off

**Comment by wangzhilong** [comment]
@CodingCat, very thank you for answering so many questions to me.

I'm not familiar with xgboost's code  implementation. i read `trainWithRDD`

1. separate training set into multiple partition, will get multiple boosters
`val boosters = buildDistributedBoosters(trainingData, overriddenParams,
        tracker.getWorkerEnvs, nWorkers, round, obj, eval, useExternalMemory, missing,
        baseMargin)`

2.  select first model as result model.
`val xgboostModel = XGBoostModel(distributedBoosters.first(), isClassificationTask)`

I think the partitions put the split point together through rabbit in each round, then select the best split point.  Is my thinking right?

**Comment by CodingCat** [comment]
yes, for each round, the model is synced

**Comment by sgatamex** [comment]
@CodingCat : There is very little documentation of inner working of distributed xgboost esp. how model is being trained and synchronized .. one has to go through various questions asked in github to understand the inner workings of distributed xgboost . Could you help to explain the tree building process in distributed xgboost using spark? 

That wud be grate help.

**Comment by superbobry** [comment]
There're some bits and pieces [here](http://learningsys.org/papers/LearningSys_2015_paper_32.pdf) and [there](https://pdfs.semanticscholar.org/cc42/7b070f214ad11f4b8e7e4e0f0a5bfa9d55bf.pdf), but no good in-depth (and up to date) description, sadly.

**Comment by CodingCat** [comment]
there is some intro in https://spark-summit.org/2017/events/building-a-unified-data-pipeline-with-apache-spark-and-xgboost/
